File: bindings/ruby/test/portable_parity_test.rb
# frozen_string_literal: true

require 'minitest/autorun'
require 'prolly'

class PortableParityTest < Minitest::Test
  def test_versioned_indexed_and_proximity_maps_use_portable_api
    Prolly::Engine.memory.use do |engine|
      versioned = engine.versioned_map('users'.b)
      versioned.initialize_map
      versioned.put('u1'.b, 'Ada'.b)
      assert_equal 'Ada'.b, versioned.get('u1'.b)

      registry = engine.index_registry
      registry.register(
        'by_team'.b, 1, 'team-v1', Prolly::IndexProjectionRecord::ALL,
        ->(_primary_key, source_value) { [[source_value, nil]] }
      )
      indexed = engine.indexed_map('members'.b, registry)
      indexed.put('u1'.b, 'red'.b)
      indexed.ensure_index('by_team'.b)
      rows = indexed.snapshot.index('by_team'.b).records('red'.b)
      assert_equal ['u1'.b], rows.map(&:primary_key)
      assert_equal 'members'.b, indexed.id
      applied = indexed.apply([
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'u2'.b, value: 'red'.b)
      ])
      conditional = indexed.apply_if(applied.source_version, [
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'u3'.b, value: 'blue'.b)
      ])
      refute_nil conditional.current
      assert_equal 2, indexed.snapshot_at(applied.source_version).index('by_team'.b).exact_page('red'.b).matches.size
      current_indexed = indexed.snapshot
      assert_equal current_indexed.id, indexed.snapshot_by_id(current_indexed.id).id
      assert_equal 'by_team'.b, current_indexed.index('by_team'.b).name

      proximity = engine.build_proximity(
        dimensions: 2,
        records: [Prolly::ProximityRecord.new(key: 'a'.b, vector: [0.0, 0.0], value: 'alpha'.b)]
      )
      assert_equal 'a'.b, proximity.search_exact([0.1, 0.1], 1).neighbors.first.key
      viewed = proximity.search_view([0.1, 0.1], 1) { |neighbors| neighbors.first.key.bytes }
      assert_equal 'a'.b, viewed
    end
  end

  def test_future_copies_inputs_before_dispatch
    Prolly::Engine.memory.use do |engine|
      versioned = engine.versioned_map('async'.b)
      versioned.initialize_map
      key = 'k'.b
      future = versioned.put_async(key, 'v'.b)
      key.replace('x'.b)
      future.value
      assert_equal 'v'.b, versioned.get('k'.b)
    end
  end

  def test_versioned_snapshot_lifecycle
    Prolly::Engine.memory.use do |engine|
      versioned = engine.versioned_map('versioned-lifecycle'.b)
      assert_equal 'versioned-lifecycle'.b, versioned.id
      refute versioned.initialized?
      initial = versioned.initialize_map
      assert versioned.initialized?
      assert_equal initial.id, versioned.head_id
      first = versioned.put('k'.b, 'v1'.b)
      versioned.put('k'.b, 'v2'.b)
      assert_equal versioned.head.id, versioned.head_id
      assert_equal first.id, versioned.version(first.id).id
      assert_operator versioned.versions.size, :>=, 3
      historical = versioned.snapshot_at(first.id)
      assert_equal first.id, historical.id
      assert_equal first.id, historical.version.id
      assert_equal 'v1'.b, historical.get('k'.b)
    end
  end

  def test_versioned_snapshots_expose_ordered_navigation_and_bounded_pages
    Prolly::Engine.memory.use do |engine|
      versioned = engine.versioned_map('versioned-ordered'.b)
      versioned.initialize_map
      versioned.apply([
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'a'.b, value: 'one'.b),
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'ab'.b, value: 'two'.b),
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'b'.b, value: 'three'.b),
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'c'.b, value: 'four'.b)
      ])
      snapshot = versioned.snapshot
      assert snapshot.contains?('ab'.b)
      assert_equal ['one'.b, nil], snapshot.get_many(['a'.b, 'missing'.b])
      assert_equal 'a'.b, snapshot.first_entry.key
      assert_equal 'c'.b, snapshot.last_entry.key
      assert_equal 'ab'.b, snapshot.lower_bound('aa'.b).key
      assert_equal 'b'.b, snapshot.upper_bound('ab'.b).key
      assert_equal ['a'.b, 'ab'.b], snapshot.prefix('a'.b).map(&:key)
      assert_equal ['ab'.b, 'b'.b], snapshot.range('ab'.b, 'c'.b).map(&:key)
      prefix_page = snapshot.prefix_page('a'.b, nil, 1)
      assert_equal ['a'.b], prefix_page.entries.map(&:key)
      refute_nil prefix_page.next_cursor
      first = snapshot.range_page(nil, 'c'.b, 2)
      assert_equal ['a'.b, 'ab'.b], first.entries.map(&:key)
      second = snapshot.range_page(first.next_cursor, 'c'.b, 2)
      assert_equal ['b'.b], second.entries.map(&:key)
      assert_equal ['c'.b, 'b'.b], snapshot.reverse_page(nil, 'a'.b, 2).entries.map(&:key)
      assert_equal ['ab'.b, 'a'.b], snapshot.prefix_reverse_page('a'.b, nil, 2).entries.map(&:key)
    end
  end

  def test_versioned_batch_cas_and_pinned_point_reads
    Prolly::Engine.memory.use do |engine|
      versioned = engine.versioned_map('versioned-cas'.b)
      versioned.initialize_map
      first = versioned.apply([
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'a'.b, value: 'one'.b),
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'b'.b, value: 'two'.b)
      ])
      assert versioned.contains?('a'.b)
      assert_equal ['one'.b, nil], versioned.get_many(['a'.b, 'missing'.b])
      applied = versioned.put_if(first.id, 'a'.b, 'updated'.b)
      assert_equal Prolly::MapUpdateKind::APPLIED, applied.kind
      assert_equal Prolly::MapUpdateKind::CONFLICT, versioned.delete_if(first.id, 'b'.b).kind
      assert_equal ['one'.b, 'two'.b], versioned.get_many_at(first.id, ['a'.b, 'b'.b])
      assert_equal 'one'.b, versioned.get_at(first.id, 'a'.b)
      result = versioned.apply_if(applied.current.id, [
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::DELETE, key: 'b'.b, value: nil)
      ])
      assert_equal Prolly::MapUpdateKind::APPLIED, result.kind
    end
  end

  def test_versioned_backup_restore_and_retention
    Prolly::Engine.memory.use do |source_engine|
      Prolly::Engine.memory.use do |target_engine|
        source = source_engine.versioned_map('versioned-backup'.b)
        source.initialize_map
        source.put('k'.b, 'v1'.b)
        source.put('k'.b, 'v2'.b)
        target = target_engine.versioned_map('versioned-backup'.b)
        restored = target.restore_backup(source.backup)
        assert_equal source.head_id, restored.id
        assert_equal 'v2'.b, target.get('k'.b)
        pruned = source.keep_last(1)
        refute_empty pruned.retained
        refute_empty pruned.removed
      end
    end
  end

  def test_proofs_sessions_and_maintenance_are_application_facing
    Prolly::Engine.memory.use do |engine|
      versioned = engine.versioned_map('proofs'.b)
      versioned.initialize_map
      versioned.put('k'.b, 'v'.b)
      versioned.put('ka'.b, 'v2'.b)
      snapshot = versioned.snapshot
      verified = Prolly.verify_key_proof(snapshot.prove_key('k'.b))
      assert verified.valid
      assert_equal 'v'.b, verified.value
      multi = Prolly.verify_multi_key_proof(snapshot.prove_keys(['k'.b, 'missing'.b]))
      assert_equal [true, false], multi.results.map(&:exists)
      ranged = Prolly.verify_range_proof(snapshot.prove_range('k'.b, 'l'.b))
      assert_equal ['k'.b, 'ka'.b], ranged.entries.map(&:key)
      prefixed = Prolly.verify_range_proof(snapshot.prove_prefix('k'.b))
      assert_equal ['k'.b, 'ka'.b], prefixed.entries.map(&:key)
      proved_page = snapshot.prove_range_page(nil, 'l'.b, 1)
      assert Prolly.verify_range_page_proof(proved_page.proof).valid
      assert_equal ['k'.b], proved_page.page.entries.map(&:key)
      assert_equal 2, snapshot.stats.total_key_value_pairs
      refute_empty snapshot.export.nodes
      snapshot.read.use do |session|
        assert_equal 'v'.b, session.get('k'.b)
        escaped = nil
        seen = []
        outcome = session.scan_range_view('k'.b, 'l'.b) do |entry|
          escaped ||= entry.key
          seen << "#{entry.key.bytes}=#{entry.value.bytes}"
          true
        end
        assert_equal 2, outcome.visited
        refute outcome.stopped
        assert_equal ['k=v', 'ka=v2'], seen
        assert_raises(RuntimeError) { escaped.bytes }
        stopped = session.scan_range_view('k'.b, 'l'.b) { false }
        assert_equal 1, stopped.visited
        assert stopped.stopped
      end
      assert_operator versioned.verify_catalog.version_count, :>=, 2
      refute_empty versioned.backup
      refute_empty versioned.plan_gc.reachability.live_cids

      registry = engine.index_registry
      registry.register(
        'by_value'.b, 1, 'value-v1', Prolly::IndexProjectionRecord::ALL,
        ->(_key, value) { [[value, nil]] }
      )
      indexed = engine.indexed_map('indexed-maintenance'.b, registry)
      version = indexed.put('k'.b, 'term'.b)
      indexed.ensure_index('by_value'.b)
      assert indexed.verify_index('by_value'.b, version.source_version).valid
      assert_operator indexed.metrics.build_attempts, :>=, 1
      refute_empty indexed.export_current
      refute_empty indexed.keep_last(1).retained_source_versions

      proximity = engine.build_proximity(
        dimensions: 2,
        records: [Prolly::ProximityRecord.new(key: 'p'.b, vector: [0.0, 0.0], value: 'payload'.b)]
      )
      membership = Prolly.verify_proximity_membership_proof(
        proximity.prove_membership('p'.b), proximity.descriptor
      )
      assert_equal 'payload'.b, membership.record.value
      assert_equal 1, proximity.verify.record_count
      assert_equal 1, proximity.count
      assert proximity.contains?('p'.b)
      assert_equal 2, proximity.config.dimensions
      structure = Prolly.verify_proximity_structure_proof(
        proximity.prove_structure, proximity.descriptor, Prolly.default_content_graph_limits
      )
      assert_equal 1, structure.summary.record_count
      mutated, stats = proximity.mutate([
        Prolly::ProximityMutationRecord.new(key: 'q'.b, vector: [1.0, 1.0], value: 'second'.b)
      ])
      assert_equal 2, mutated.count
      assert_operator stats.records_rebuilt, :>=, 1
      proximity.read.use do |retained|
        assert_equal 'p'.b, retained.search_exact([0.0, 0.0], 1).neighbors.first.key
        assert_equal 'p'.b, retained.search_view([0.0, 0.0], 1) { |rows| rows.first.key.to_s }
      end
      proximity.prove_search_exact([0.0, 0.0], 1).use do |proof|
        verified_search = proof.verify(proximity.descriptor)
        assert_equal 'p'.b, verified_search.result.neighbors.first.key
        assert_operator verified_search.replayed_events, :>, 0
      end
    end
  end

  def test_versioned_comparison_pins_versions_and_pages_diffs
    Prolly::Engine.memory.use do |engine|
      map = engine.versioned_map('comparison'.b)
      base = map.initialize_map
      target = map.put('k'.b, 'v'.b)
      comparison = map.compare(base.id, target.id)
      assert_equal base.id, comparison.base.id
      assert_equal target.id, comparison.target.id
      assert_equal ['k'.b], comparison.diff.map(&:key)
      assert_equal ['k'.b], comparison.diff_page(nil, nil, 1).diffs.map(&:key)
      comparison.close
    end
  end


  def test_versioned_subscription_resumes_and_polls_owned_diffs
    Prolly::Engine.memory.use do |engine|
      map = engine.versioned_map('subscription'.b)
      initial = map.initialize_map
      subscription = map.subscribe
      assert_equal initial.id, subscription.last_seen
      assert_nil subscription.poll
      current = map.put('k'.b, 'v'.b)
      event = subscription.poll
      assert_equal initial.id, event.previous
      assert_equal current.id, event.current.id
      assert_equal ['k'.b], event.diffs.map(&:key)
      assert_equal current.id, subscription.last_seen
      subscription.close
    end
  end
end

File: bindings/node/test/portable-parity.test.ts
import test from "node:test";
import assert from "node:assert/strict";

import { Engine, exactSearch } from "../src/index.ts";

const bytes = (value: string): Buffer => Buffer.from(value);

test("portable versioned, indexed, and proximity maps stay native", async () => {
  const engine = await Engine.memory();
  try {
    const versioned = engine.versionedMap(bytes("users"));
    await versioned.initialize();
    await versioned.put(bytes("u1"), bytes("Ada"));
    assert.equal(Buffer.from(await versioned.get(bytes("u1")) ?? []).toString(), "Ada");

    const registry = engine.indexRegistry();
    registry.register({
      name: bytes("by_team"),
      generation: 1n,
      extractorId: "team-v1",
      projection: "all",
      extract: (_key, source) => [{ term: Buffer.from(source) }],
    });
    const indexed = engine.indexedMap(bytes("members"), registry);
    await indexed.put(bytes("u1"), bytes("red"));
    await indexed.put(bytes("u2"), bytes("red"));
    await indexed.ensureIndex(bytes("by_team"));
    const index = (await indexed.snapshot()).index(bytes("by_team"));
    const records = await index.records(bytes("red"));
    assert.equal(Buffer.from(records[0].primaryKey).toString(), "u1");
    assert.equal(Buffer.from(records[0].sourceValue).toString(), "red");
    const pageKeys: string[] = [];
    for await (const page of index.exactPages(bytes("red"), { pageSize: 1 })) {
      pageKeys.push(...page.map((row) => Buffer.from(row.primaryKey).toString()));
    }
    assert.deepEqual(pageKeys, ["u1", "u2"]);

    let escaped: Uint8Array | undefined;
    await index.exactView(bytes("red"), (row) => {
      escaped = row.primaryKey;
      assert.equal(Buffer.from(row.primaryKey).toString(), "u1");
      return false;
    });
    assert.throws(() => escaped?.byteLength, /expired/i);

    const proximity = await engine.buildProximity(2, [
      { key: bytes("a"), vector: new Float32Array([0, 0]), value: bytes("alpha") },
    ]);
    const proximitySession = proximity.read();
    assert.equal(proximitySession.contains(bytes("a")), true);
    assert.equal(Buffer.from(proximitySession.get(bytes("a"))?.value ?? []).toString(), "alpha");
    assert.ok(proximitySession.fastHandle() > 0n);
    const result = await proximitySession.search(exactSearch(new Float32Array([0.1, 0.1]), 1));
    assert.equal(Buffer.from(result.neighbors[0].key).toString(), "a");
    proximitySession.close();
  } finally {
    engine.close();
  }
});

test("portable promises honor AbortSignal and own inputs", async () => {
  const engine = await Engine.memory();
  try {
    const map = engine.versionedMap(bytes("async"));
    await map.initialize();
    const key = bytes("original-key");
    const value = bytes("original-value");
    const pending = map.put(key, value);
    key.fill(0);
    value.fill(0);
    await pending;
    assert.equal(Buffer.from(await map.get(bytes("original-key")) ?? []).toString(), "original-value");

    const controller = new AbortController();
    controller.abort();
    await assert.rejects(map.get(bytes("original-key"), controller.signal), /abort/i);
  } finally {
    engine.close();
  }
});

test("versioned maps expose identity and historical snapshot lifecycle", async () => {
  const engine = await Engine.memory();
  try {
    const map = engine.versionedMap(bytes("versioned-lifecycle"));
    assert.equal(Buffer.from(map.id()).toString(), "versioned-lifecycle");
    assert.equal(await map.isInitialized(), false);
    const initial = await map.initialize();
    assert.equal(await map.isInitialized(), true);
    assert.deepEqual(await map.headId(), initial.id);
    const first = await map.put(bytes("k"), bytes("v1"));
    await map.put(bytes("k"), bytes("v2"));
    assert.deepEqual((await map.head())?.id, (await map.headId()));
    assert.deepEqual((await map.version(first.id))?.id, first.id);
    assert.ok((await map.versions()).length >= 3);
    const historical = await map.snapshotAt(first.id);
    assert.ok(historical);
    assert.deepEqual(historical.id(), first.id);
    assert.deepEqual(historical.version().id, first.id);
    assert.equal(Buffer.from(await historical.get(bytes("k")) ?? []).toString(), "v1");
  } finally {
    engine.close();
  }
});

test("proofs, retained sessions, and maintenance stay native", async () => {
  const engine = await Engine.memory();
  try {
    const versioned = engine.versionedMap(bytes("proofs"));
    await versioned.initialize();
    await versioned.put(bytes("k"), bytes("v"));
    const snapshot = await versioned.snapshot();
    assert.ok(snapshot);
    const proof = snapshot.proveKey(bytes("k"));
    assert.equal(Buffer.from(proof.verify().value ?? []).toString(), "v");
    assert.equal(snapshot.stats().itemCount, 1n);
    assert.ok(snapshot.exportSummary().itemCount > 0n);
    const session = snapshot.read();
    assert.equal(Buffer.from(session.get(bytes("k")) ?? []).toString(), "v");
    assert.ok((await versioned.verifyCatalog()).itemCount >= 2n);
    assert.ok((await versioned.backup()).byteLength > 0);
    assert.ok((await versioned.planGc()).itemCount > 0n);

    const registry = engine.indexRegistry();
    registry.register({
      name: bytes("by_value"), generation: 1n, extractorId: "value-v1", projection: "all",
      extract: (_key, value) => [{ term: Buffer.from(value) }],
    });
    const indexed = engine.indexedMap(bytes("indexed-maintenance"), registry);
    const version = await indexed.put(bytes("k"), bytes("term"));
    await indexed.ensureIndex(bytes("by_value"));
    assert.equal(indexed.verifyIndex(bytes("by_value"), version.sourceVersion).valid, true);
    assert.ok(indexed.metrics().buildAttempts >= 1n);
    assert.ok(indexed.exportCurrent().byteLength > 0);
    assert.ok(indexed.keepLast(1n).retainedSourceVersions.length >= 1);

    const proximity = await engine.buildProximity(2, [
      { key: bytes("p"), vector: new Float32Array([0, 0]), value: bytes("payload") },
    ]);
    const membership = proximity.proveMembership(bytes("p")).verify(proximity.descriptor());
    assert.equal(Buffer.from(membership.value ?? []).toString(), "payload");
    assert.equal(proximity.verify().recordCount, 1n);
    assert.equal(proximity.count(), 1n);
    assert.equal(proximity.contains(bytes("p")), true);
    assert.equal(proximity.config().dimensions, 2);
    const structural = proximity.proveStructure().verify(proximity.descriptor());
    assert.equal(structural.summary.recordCount, 1n);
    const mutation = proximity.mutate([
      { key: bytes("q"), vector: new Float32Array([1, 1]), value: bytes("second") },
    ]);
    assert.equal(mutation.map.count(), 2n);
    assert.ok(mutation.stats.recordsRebuilt >= 1n);
    const searchProof = proximity.proveSearch(exactSearch(new Float32Array([0, 0]), 1));
    const verifiedSearch = searchProof.verify(proximity.descriptor());
    assert.equal(Buffer.from(verifiedSearch.result.neighbors[0]!.key).toString(), "p");
    assert.ok(verifiedSearch.replayedEvents > 0n);
    searchProof.close();
  } finally {
    engine.close();
  }
});

test("indexed maps expose batch CAS and historical snapshot lifecycle", async () => {
  const engine = await Engine.memory();
  try {
    const registry = engine.indexRegistry();
    registry.register({
      name: bytes("by_value"), generation: 1n, extractorId: "value-v1", projection: "all",
      extract: (_key, value) => [{ term: Buffer.from(value) }],
    });
    const indexed = engine.indexedMap(bytes("indexed-lifecycle"), registry);
    assert.equal(Buffer.from(indexed.id()).toString(), "indexed-lifecycle");

    const first = await indexed.apply([
      { kind: "upsert", key: bytes("u1"), value: bytes("red") },
      { kind: "upsert", key: bytes("u2"), value: bytes("red") },
    ]);
    await indexed.ensureIndex(bytes("by_value"));
    const firstSnapshot = await indexed.snapshot();
    const firstSnapshotId = firstSnapshot.id();
    assert.deepEqual(firstSnapshotId.sourceVersion, first.sourceVersion);

    const applied = await indexed.applyIf(first.sourceVersion, [
      { kind: "upsert", key: bytes("u3"), value: bytes("blue") },
    ]);
    assert.equal(applied.kind, "applied");
    assert.ok(applied.current);
    const conflict = await indexed.applyIf(first.sourceVersion, [
      { kind: "delete", key: bytes("u1") },
    ]);
    assert.equal(conflict.kind, "conflict");

    const historical = await indexed.snapshotAt(first.sourceVersion);
    assert.deepEqual(historical.id().sourceVersion, firstSnapshotId.sourceVersion);
    const reopened = await indexed.snapshotById(firstSnapshotId);
    assert.deepEqual(reopened.id(), firstSnapshotId);
  } finally {
    engine.close();
  }
});

test("indexed maintenance returns complete portable records", async () => {
  const engine = await Engine.memory();
  try {
    const registry = engine.indexRegistry();
    registry.register({
      name: bytes("by_value"), generation: 1n, extractorId: "value-v1", projection: "all",
      extract: (_key, value) => [{ term: Buffer.from(value) }],
    });
    const indexed = engine.indexedMap(bytes("indexed-records"), registry);
    const version = await indexed.put(bytes("u1"), bytes("red"));
    await indexed.ensureIndex(bytes("by_value"));

    const health = indexed.health();
    assert.equal(Buffer.from(health.sourceMapId).toString(), "indexed-records");
    assert.equal(health.activeIndexes.length, 1);
    assert.equal(health.activeIndexes[0]!.generation, 1n);
    const verification = indexed.verifyIndex(bytes("by_value"), version.sourceVersion);
    assert.equal(verification.valid, true);
    assert.equal(verification.canonical, true);
    assert.equal(indexed.verifyAll(version.sourceVersion).length, 1);
    assert.equal(indexed.repairIndex(bytes("by_value"), version.sourceVersion).valid, true);
    assert.ok(indexed.metrics().buildAttempts >= 1n);

    const bundle = indexed.exportCurrent();
    const next = await indexed.put(bytes("u2"), bytes("blue"));
    const imported = await indexed.importCurrent(bundle, next.sourceVersion);
    assert.deepEqual(imported.sourceVersion, version.sourceVersion);
    const retained = indexed.keepLast(1n);
    assert.ok(retained.retainedSourceVersions.length >= 1);
    await indexed.deactivateIndex(bytes("by_value"));
    assert.equal(indexed.health().activeIndexes.length, 0);
  } finally {
    engine.close();
  }
});

test("secondary indexes expose identity and every bounded page direction", async () => {
  const engine = await Engine.memory();
  try {
    const registry = engine.indexRegistry();
    registry.register({
      name: bytes("by_value"), generation: 1n, extractorId: "value-v1", projection: "all",
      extract: (_key, value) => [{ term: Buffer.from(value) }],
    });
    const indexed = engine.indexedMap(bytes("indexed-pages"), registry);
    await indexed.apply([
      { kind: "upsert", key: bytes("u1"), value: bytes("red") },
      { kind: "upsert", key: bytes("u2"), value: bytes("red") },
      { kind: "upsert", key: bytes("u3"), value: bytes("rose") },
    ]);
    await indexed.ensureIndex(bytes("by_value"));
    const index = (await indexed.snapshot()).index(bytes("by_value"));
    assert.equal(Buffer.from(index.name()).toString(), "by_value");

    const keys = async (pages: AsyncIterable<readonly { primaryKey: Uint8Array }[]>) => {
      const result: string[] = [];
      for await (const page of pages) {
        result.push(...page.map((row) => Buffer.from(row.primaryKey).toString()));
      }
      return result;
    };
    assert.deepEqual(await keys(index.exactPages(bytes("red"), { pageSize: 1 })), ["u1", "u2"]);
    assert.deepEqual(await keys(index.exactReversePages(bytes("red"), { pageSize: 1 })), ["u2", "u1"]);
    assert.deepEqual(await keys(index.prefixPages(bytes("r"), { pageSize: 1 })), ["u1", "u2", "u3"]);
    assert.deepEqual(await keys(index.prefixReversePages(bytes("r"), { pageSize: 1 })), ["u3", "u2", "u1"]);
    assert.deepEqual(await keys(index.rangePages(bytes("red"), bytes("s"), { pageSize: 1 })), ["u1", "u2", "u3"]);
    assert.deepEqual(await keys(index.rangeReversePages(bytes("red"), bytes("s"), { pageSize: 1 })), ["u3", "u2", "u1"]);
    assert.equal(Buffer.from((await index.exactPage(bytes("red"), undefined, 1n)).matches[0]!.primaryKey).toString(), "u1");
    assert.equal(Buffer.from((await index.exactReversePage(bytes("red"), undefined, 1n)).matches[0]!.primaryKey).toString(), "u2");
    assert.equal(Buffer.from((await index.prefixPage(bytes("r"), undefined, 1n)).matches[0]!.primaryKey).toString(), "u1");
    assert.equal(Buffer.from((await index.prefixReversePage(bytes("r"), undefined, 1n)).matches[0]!.primaryKey).toString(), "u3");
    assert.equal(Buffer.from((await index.rangePage(bytes("red"), bytes("s"), undefined, 1n)).matches[0]!.primaryKey).toString(), "u1");
    assert.equal(Buffer.from((await index.rangeReversePage(bytes("red"), bytes("s"), undefined, 1n)).matches[0]!.primaryKey).toString(), "u3");
  } finally {
    engine.close();
  }
});
